Sources/MongoSwift/MongoCursor.swift

```diff
@@ -99,21 +139,40 @@ public class MongoCursor<T: Codable>: Sequence, IteratorProtocol {
             return extractMongoError(error: error, reply: reply)
         }
 
-        // Otherwise, the only feasible error is that the user tried to advance a dead cursor, which is a logic error.
-        // We will still parse the mongoc error to cover all cases.
+        // The only feasible error here is that we tried to advance a dead mongoc cursor. Due to our cursor-closing
+        // logic in `next()` that should never happen, but parse the error anyway just in case we end up here.
         return extractMongoError(error: error)
     }
 
     /// Returns the next `Document` in this cursor, or nil. Once this function returns `nil`, the caller should use
     /// the `.error` property to check for errors.
     public func next() -> T? {
+        // We already closed the mongoc cursor, either because we reached the end or encountered an error.
+        guard case let .open(cursor, conn, _, session) = self.state else {
+            self.error = ClosedCursorError
+            return nil
+        }
+
         do {
             let operation = NextOperation(cursor: self)
-            let out = try operation.execute(using: self._connection, session: self._session)
-            self.swiftError = nil
+            guard let out = try operation.execute(using: conn, session: session) else {
```

Comment:
something to note about this design: for a (non-tailable) cursor with `x` documents, we won't close the cursor until the `x + 1`th call to `next()`, where `mongoc_cursor_next` returns `nil`.
we could consider calling `mongoc_cursor_next` and preloading the next document each time, but I wasn't sure if that was worth doing...  we generally expect users to iterate via for loop, right? in which case they will always make all `x + 1` calls to `next()` since that calls `next()` until it returns `nil`.
(note that for non-tailable `mongoc_cursor_more` won't help, as it will always return `true` until a call to `mongoc_cursor_next` has returned `false`.)

in the tailable case,  we *could* preemptively call `mongoc_cursor_more` after each non-nil `out` value and see if there is potentially anything else or if we just retrieved the last doc. (right now we only do this after a nil `out`), and if we are at the end immediately close the cursor but allow the user to call `next()` one more time without triggering a `ClosedCursorError`.

all of this seemed a bit overly complicated to me but if anyone feels like either of those changes would be worth making let me know.

Reply:
Would this code trigger the issue:
```
let singleDocFromCursor = try cursor.next()
.. do something with the single document from the response, but don't keep iterating ..
```
?

Even in that case it might not be an issue because the cursor will likely fall out of scope (where we will clean it up implicitly). I think this behavior is fine. If we had more of a java iterator approach, then the `hasNext` would need to cache a document like one of your proposals above:
```
while (cursor.hasNext()) {
   doSomethingWith(cursor.next())
}
```

Reply:
yes, it would trigger it unless if no documents were actually returned in the cursor. if the cursor has one or more documents, then it will stay open until it goes out of scope.